# Problem Statement
Output variable -> y
y -> Whether the client has subscribed a term deposit or not 
Binomial ("yes" or "no")
Attribute information For bank dataset

   Input variables:
   # bank client data:
   1 - age (numeric)
   2 - job : type of job (categorical: "admin.","unknown","unemployed","management","housemaid","entrepreneur","student",
                                       "blue-collar","self-employed","retired","technician","services") 
   3 - marital : marital status (categorical: "married","divorced","single"; note: "divorced" means divorced or widowed)
   4 - education (categorical: "unknown","secondary","primary","tertiary")
   5 - default: has credit in default? (binary: "yes","no")
   6 - balance: average yearly balance, in euros (numeric) 
   7 - housing: has housing loan? (binary: "yes","no")
   8 - loan: has personal loan? (binary: "yes","no")
   # related with the last contact of the current campaign:
   9 - contact: contact communication type (categorical: "unknown","telephone","cellular") 
  10 - day: last contact day of the month (numeric)
  11 - month: last contact month of year (categorical: "jan", "feb", "mar", ..., "nov", "dec")
  12 - duration: last contact duration, in seconds (numeric)
   # other attributes:
  13 - campaign: number of contacts performed during this campaign and for this client (numeric, includes last contact)
  14 - pdays: number of days that passed by after the client was last contacted from a previous campaign (numeric, -1 means client was not previously contacted)
  15 - previous: number of contacts performed before this campaign and for this client (numeric)
  16 - poutcome: outcome of the previous marketing campaign (categorical: "unknown","other","failure","success")

  Output variable (desired target):
  17 - y - has the client subscribed a term deposit? (binary: "yes","no")


In [1]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
import warnings
warnings.filterwarnings("ignore")

In [2]:
data=pd.read_csv('bank-full.csv',sep=';')   

In [3]:
data2=data
data1=data

In [278]:
#data1.head(10).reset_index()
data1

age           job   marital  education default  balance housing loan  \
0       58    management   married   tertiary      no     2143     yes   no   
1       44    technician    single  secondary      no       29     yes   no   
2       33  entrepreneur   married  secondary      no        2     yes  yes   
3       47   blue-collar   married    unknown      no     1506     yes   no   
4       33       unknown    single    unknown      no        1      no   no   
...    ...           ...       ...        ...     ...      ...     ...  ...   
45206   51    technician   married   tertiary      no      825      no   no   
45207   71       retired  divorced    primary      no     1729      no   no   
45208   72       retired   married  secondary      no     5715      no   no   
45209   57   blue-collar   married  secondary      no      668      no   no   
45210   37  entrepreneur   married  secondary      no     2971      no   no   

         contact  day month  duration  campaign  pdays  previous poutcome    y  
0        unknown    5   may       261         1     -1         0  unknown   no  
1        unknown    5   may       151         1     -1         0  unknown   no  
2        unknown    5   may        76         1     -1         0  unknown   no  
3        unknown    5   may        92         1     -1         0  unknown   no  
4        unknown    5   may       198         1     -1         0  unknown   no  
...          ...  ...   ...       ...       ...    ...       ...      ...  ...  
45206   cellular   17   nov       977         3     -1         0  unknown  yes  
45207   cellular   17   nov       456         2     -1         0  unknown  yes  
45208   cellular   17   nov      1127         5    184         3  success  yes  
45209  telephone   17   nov       508         4     -1         0  unknown   no  
45210   cellular   17   nov       361         2    188        11    other   no  

[45211 rows x 17 columns]

In [279]:
#data1.columns=data1.iloc[0]
#data1=data1.drop(0)             #assigning first row as column 

In [280]:
data1.head(10)

age           job   marital  education default  balance housing loan  \
0   58    management   married   tertiary      no     2143     yes   no   
1   44    technician    single  secondary      no       29     yes   no   
2   33  entrepreneur   married  secondary      no        2     yes  yes   
3   47   blue-collar   married    unknown      no     1506     yes   no   
4   33       unknown    single    unknown      no        1      no   no   
5   35    management   married   tertiary      no      231     yes   no   
6   28    management    single   tertiary      no      447     yes  yes   
7   42  entrepreneur  divorced   tertiary     yes        2     yes   no   
8   58       retired   married    primary      no      121     yes   no   
9   43    technician    single  secondary      no      593     yes   no   

   contact  day month  duration  campaign  pdays  previous poutcome   y  
0  unknown    5   may       261         1     -1         0  unknown  no  
1  unknown    5   may       151         1     -1         0  unknown  no  
2  unknown    5   may        76         1     -1         0  unknown  no  
3  unknown    5   may        92         1     -1         0  unknown  no  
4  unknown    5   may       198         1     -1         0  unknown  no  
5  unknown    5   may       139         1     -1         0  unknown  no  
6  unknown    5   may       217         1     -1         0  unknown  no  
7  unknown    5   may       380         1     -1         0  unknown  no  
8  unknown    5   may        50         1     -1         0  unknown  no  
9  unknown    5   may        55         1     -1         0  unknown  no

In [257]:
data.shape

(45212, 17)

In [258]:
data1.isna().sum()

0
age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
y            0
dtype: int64

In [259]:
data1.loc[data1['housing']=='yes','housing']=1
data1.loc[data1['housing']=='no','housing']=0

In [260]:
data1.default

1        no
2        no
3        no
4        no
5        no
         ..
45207    no
45208    no
45209    no
45210    no
45211    no
Name: default, Length: 45211, dtype: object

In [261]:
data1.loc[data1['default']=='yes','default']=1
data1.loc[data1['default']=='no','default']=0

In [262]:
data1.loc[data1['loan']=='yes','loan']=1
data1.loc[data1['loan']=='no','loan']=0

In [263]:
data1.loc[data1['y']=='yes','y']=1
data1.loc[data1['y']=='no','y']=0

In [264]:
#data1.drop(['job','marital','education','contact','poutcome'],axis=1,inplace=True)


In [265]:
data1.head(4)

0 age           job  marital  education default balance housing loan  contact  \
1  58    management  married   tertiary       0    2143       1    0  unknown   
2  44    technician   single  secondary       0      29       1    0  unknown   
3  33  entrepreneur  married  secondary       0       2       1    1  unknown   
4  47   blue-collar  married    unknown       0    1506       1    0  unknown   

0 day month duration campaign pdays previous poutcome  y  
1   5   may      261        1    -1        0  unknown  0  
2   5   may      151        1    -1        0  unknown  0  
3   5   may       76        1    -1        0  unknown  0  
4   5   may       92        1    -1        0  unknown  0

In [266]:
data1.month.value_counts()

may    13766
jul     6895
aug     6247
jun     5341
nov     3970
apr     2932
feb     2649
jan     1403
oct      738
sep      579
mar      477
dec      214
Name: month, dtype: int64

In [267]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
data1["marital1"]=le.fit_transform(data1.marital)#married =1,single=2,divorced=0

In [268]:
data1.marital.value_counts()

married     27214
single      12790
divorced     5207
Name: marital, dtype: int64

In [269]:
data1.marital1

1        1
2        2
3        1
4        1
5        2
        ..
45207    1
45208    0
45209    1
45210    1
45211    1
Name: marital1, Length: 45211, dtype: int64

In [270]:
data1.head(5)

0 age           job  marital  education default balance housing loan  contact  \
1  58    management  married   tertiary       0    2143       1    0  unknown   
2  44    technician   single  secondary       0      29       1    0  unknown   
3  33  entrepreneur  married  secondary       0       2       1    1  unknown   
4  47   blue-collar  married    unknown       0    1506       1    0  unknown   
5  33       unknown   single    unknown       0       1       0    0  unknown   

0 day month duration campaign pdays previous poutcome  y  marital1  
1   5   may      261        1    -1        0  unknown  0         1  
2   5   may      151        1    -1        0  unknown  0         2  
3   5   may       76        1    -1        0  unknown  0         1  
4   5   may       92        1    -1        0  unknown  0         1  
5   5   may      198        1    -1        0  unknown  0         2

In [271]:
le=LabelEncoder()
data1["poutcome1"]=le.fit_transform(data1.poutcome)

In [272]:
data1.poutcome.value_counts()#-unknown-3 success-2 failure-0 other-1

unknown    36959
failure     4901
other       1840
success     1511
Name: poutcome, dtype: int64

In [273]:
data1.tail(10)

0     age           job   marital  education default balance housing loan  \
45202  53    management   married   tertiary       0     583       0    0   
45203  34        admin.    single  secondary       0     557       0    0   
45204  23       student    single   tertiary       0     113       0    0   
45205  73       retired   married  secondary       0    2850       0    0   
45206  25    technician    single  secondary       0     505       0    1   
45207  51    technician   married   tertiary       0     825       0    0   
45208  71       retired  divorced    primary       0    1729       0    0   
45209  72       retired   married  secondary       0    5715       0    0   
45210  57   blue-collar   married  secondary       0     668       0    0   
45211  37  entrepreneur   married  secondary       0    2971       0    0   

0        contact day month duration campaign pdays previous poutcome  y  \
45202   cellular  17   nov      226        1   184        4  success  1   
45203   cellular  17   nov      224        1    -1        0  unknown  1   
45204   cellular  17   nov      266        1    -1        0  unknown  1   
45205   cellular  17   nov      300        1    40        8  failure  1   
45206   cellular  17   nov      386        2    -1        0  unknown  1   
45207   cellular  17   nov      977        3    -1        0  unknown  1   
45208   cellular  17   nov      456        2    -1        0  unknown  1   
45209   cellular  17   nov     1127        5   184        3  success  1   
45210  telephone  17   nov      508        4    -1        0  unknown  0   
45211   cellular  17   nov      361        2   188       11    other  0   

0      marital1  poutcome1  
45202         1          2  
45203         2          3  
45204         2          3  
45205         1          0  
45206         2          3  
45207         1          3  
45208         0          3  
45209         1          2  
45210         1          3  
45211         1          1

In [274]:
data1.contact.value_counts()

cellular     29285
unknown      13020
telephone     2906
Name: contact, dtype: int64

In [275]:
lm=LabelEncoder()
data1['contact1']=lm.fit_transform(data1.contact)

In [202]:
data1.head(10)


0  age           job   marital  education default balance housing loan  \
1   58    management   married   tertiary       0    2143       1    0   
2   44    technician    single  secondary       0      29       1    0   
3   33  entrepreneur   married  secondary       0       2       1    1   
4   47   blue-collar   married    unknown       0    1506       1    0   
5   33       unknown    single    unknown       0       1       0    0   
6   35    management   married   tertiary       0     231       1    0   
7   28    management    single   tertiary       0     447       1    1   
8   42  entrepreneur  divorced   tertiary       1       2       1    0   
9   58       retired   married    primary       0     121       1    0   
10  43    technician    single  secondary       0     593       1    0   

0   contact day month duration campaign pdays previous poutcome  y  marital1  \
1   unknown   5   may      261        1    -1        0  unknown  0         1   
2   unknown   5   may      151        1    -1        0  unknown  0         2   
3   unknown   5   may       76        1    -1        0  unknown  0         1   
4   unknown   5   may       92        1    -1        0  unknown  0         1   
5   unknown   5   may      198        1    -1        0  unknown  0         2   
6   unknown   5   may      139        1    -1        0  unknown  0         1   
7   unknown   5   may      217        1    -1        0  unknown  0         2   
8   unknown   5   may      380        1    -1        0  unknown  0         0   
9   unknown   5   may       50        1    -1        0  unknown  0         1   
10  unknown   5   may       55        1    -1        0  unknown  0         2   

0   poutcome1  contact1  
1           3         2  
2           3         2  
3           3         2  
4           3         2  
5           3         2  
6           3         2  
7           3         2  
8           3         2  
9           3         2  
10          3         2

In [203]:
data1=data1.drop(['month','day'],axis=1)

In [211]:
data1=data1.drop(['marital','contact','poutcome'],axis=1)

In [214]:
data1.head(3)

0 age default balance housing loan duration campaign pdays previous  y  \
1  58       0    2143       1    0      261        1    -1        0  0   
2  44       0      29       1    0      151        1    -1        0  0   
3  33       0       2       1    1       76        1    -1        0  0   

0  marital1  poutcome1  contact1  education1 y1  
1         1          3         2           2  0  
2         2          3         2           1  0  
3         1          3         2           1  0

In [216]:
data1.head(10)

0  age default balance housing loan duration campaign pdays previous  y  \
1   58       0    2143       1    0      261        1    -1        0  0   
2   44       0      29       1    0      151        1    -1        0  0   
3   33       0       2       1    1       76        1    -1        0  0   
4   47       0    1506       1    0       92        1    -1        0  0   
5   33       0       1       0    0      198        1    -1        0  0   
6   35       0     231       1    0      139        1    -1        0  0   
7   28       0     447       1    1      217        1    -1        0  0   
8   42       1       2       1    0      380        1    -1        0  0   
9   58       0     121       1    0       50        1    -1        0  0   
10  43       0     593       1    0       55        1    -1        0  0   

0   marital1  poutcome1  contact1  education1 y1  
1          1          3         2           2  0  
2          2          3         2           1  0  
3          1          3         2           1  0  
4          1          3         2           3  0  
5          2          3         2           3  0  
6          1          3         2           2  0  
7          2          3         2           2  0  
8          0          3         2           2  0  
9          1          3         2           0  0  
10         2          3         2           1  0

In [217]:
data1['y1']=data1.y

In [218]:
data1.drop(['y'],axis=1)

0     age default balance housing loan duration campaign pdays previous  \
1      58       0    2143       1    0      261        1    -1        0   
2      44       0      29       1    0      151        1    -1        0   
3      33       0       2       1    1       76        1    -1        0   
4      47       0    1506       1    0       92        1    -1        0   
5      33       0       1       0    0      198        1    -1        0   
...    ..     ...     ...     ...  ...      ...      ...   ...      ...   
45207  51       0     825       0    0      977        3    -1        0   
45208  71       0    1729       0    0      456        2    -1        0   
45209  72       0    5715       0    0     1127        5   184        3   
45210  57       0     668       0    0      508        4    -1        0   
45211  37       0    2971       0    0      361        2   188       11   

0      marital1  poutcome1  contact1  education1 y1  
1             1          3         2           2  0  
2             2          3         2           1  0  
3             1          3         2           1  0  
4             1          3         2           3  0  
5             2          3         2           3  0  
...         ...        ...       ...         ... ..  
45207         1          3         0           2  1  
45208         0          3         0           0  1  
45209         1          2         0           1  1  
45210         1          3         1           1  0  
45211         1          1         0           1  0  

[45211 rows x 14 columns]

In [219]:
data1.head(3)

0 age default balance housing loan duration campaign pdays previous  y  \
1  58       0    2143       1    0      261        1    -1        0  0   
2  44       0      29       1    0      151        1    -1        0  0   
3  33       0       2       1    1       76        1    -1        0  0   

0  marital1  poutcome1  contact1  education1 y1  
1         1          3         2           2  0  
2         2          3         2           1  0  
3         1          3         2           1  0

In [222]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45211 entries, 1 to 45211
Data columns (total 15 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   age         45211 non-null  object
 1   default     45211 non-null  object
 2   balance     45211 non-null  object
 3   housing     45211 non-null  object
 4   loan        45211 non-null  object
 5   duration    45211 non-null  object
 6   campaign    45211 non-null  object
 7   pdays       45211 non-null  object
 8   previous    45211 non-null  object
 9   y           45211 non-null  object
 10  marital1    45211 non-null  int64 
 11  poutcome1   45211 non-null  int64 
 12  contact1    45211 non-null  int64 
 13  education1  45211 non-null  int64 
 14  y1          45211 non-null  object
dtypes: int64(4), object(11)
memory usage: 5.5+ MB


In [223]:
Y.value_counts()

0    5489
1    5289
Name: y1, dtype: int64

# Our data is imbalnced so we need to balnce the data 
(we are going to use under sampling )

In [224]:
class_count0,class_count1=data1.y.value_counts()

In [225]:
class0=data1[data1['y']==0]

In [226]:
class1=data1[data1['y']==1]

In [227]:
class0_under=class0.sample(class_count1+200)

In [228]:
data1_b=pd.concat((class1,class0_under),axis=0) #created new data frame with balanced sample data1_b

In [229]:
data1_b.y.value_counts(0)

0    5489
1    5289
Name: y, dtype: int64

In [230]:
data1_b

0     age default balance housing loan duration campaign pdays previous  y  \
84     59       0    2343       1    0     1042        1    -1        0  1   
87     56       0      45       0    0     1467        1    -1        0  1   
88     41       0    1270       1    0     1389        1    -1        0  1   
130    55       0    2476       1    0      579        1    -1        0  1   
169    54       0     184       0    0      673        2    -1        0  1   
...    ..     ...     ...     ...  ...      ...      ...   ...      ... ..   
34797  35       0     105       1    0       36        1   365        3  0   
2098   27       0     541       1    0      145        1    -1        0  0   
7085   51       0     943       1    0      620        1    -1        0  0   
4025   60       0      26       1    0      196        6    -1        0  0   
29763  38       0     308       1    0      197        1   274        1  0   

0      marital1  poutcome1  contact1  education1 y1  
84            1          3         2           1  1  
87            1          3         2           1  1  
88            1          3         2           1  1  
130           1          3         2           1  1  
169           1          3         2           2  1  
...         ...        ...       ...         ... ..  
34797         1          0         0           2  0  
2098          2          3         2           2  0  
7085          1          3         2           0  0  
4025          1          3         2           0  0  
29763         0          0         0           1  0  

[10778 rows x 15 columns]

In [231]:
data1_b=data1_b.drop(['y'],axis=1)

In [232]:
X=data1_b.iloc[:,0:14]

In [233]:
X.head(3)

0  age default balance housing loan duration campaign pdays previous  \
84  59       0    2343       1    0     1042        1    -1        0   
87  56       0      45       0    0     1467        1    -1        0   
88  41       0    1270       1    0     1389        1    -1        0   

0   marital1  poutcome1  contact1  education1 y1  
84         1          3         2           1  1  
87         1          3         2           1  1  
88         1          3         2           1  1

In [234]:
X.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10778 entries, 84 to 29763
Data columns (total 14 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   age         10778 non-null  object
 1   default     10778 non-null  object
 2   balance     10778 non-null  object
 3   housing     10778 non-null  object
 4   loan        10778 non-null  object
 5   duration    10778 non-null  object
 6   campaign    10778 non-null  object
 7   pdays       10778 non-null  object
 8   previous    10778 non-null  object
 9   marital1    10778 non-null  int64 
 10  poutcome1   10778 non-null  int64 
 11  contact1    10778 non-null  int64 
 12  education1  10778 non-null  int64 
 13  y1          10778 non-null  object
dtypes: int64(4), object(10)
memory usage: 1.2+ MB


In [235]:
Y=data1_b.iloc[:,13]

In [236]:
Y.value_counts()

0    5489
1    5289
Name: y1, dtype: int64

In [237]:
Y=Y.astype(int)

# Logistic Regression

In [238]:
#Logistic regression and fit the model
model1 = LogisticRegression()
model1.fit(X,Y)

LogisticRegression()

In [239]:
model1.intercept_

array([-0.7144468])

In [240]:
model1.coef_

array([[-8.76337914e-04, -7.78044211e-02, -6.96685155e-05,
        -2.84037857e+00, -8.34941862e-01,  4.20492254e-03,
        -1.48790000e-02, -4.68967787e-03, -1.76326169e-01,
        -3.97965538e-01, -7.97092380e-01, -3.11683430e+00,
         6.54347705e-01,  9.23567058e+00]])

In [241]:
y_predict=model1.predict(X)

In [242]:
y_predict_df=pd.DataFrame({'Actual_Y':Y,'predict_Y':model1.predict(X)})

In [243]:
y_predict_df

Actual_Y  predict_Y
84            1          1
87            1          1
88            1          1
130           1          0
169           1          1
...         ...        ...
34797         0          0
2098          0          0
7085          0          0
4025          0          0
29763         0          0

[10778 rows x 2 columns]

In [244]:
# Confusion Matrix for the model accuracy
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score as ac
confusion_matrix = confusion_matrix(Y,y_predict)
print (confusion_matrix)

[[5412   77]
 [ 161 5128]]


In [245]:
ac(Y,y_predict)

0.9779179810725552

In [246]:
#Classification report
from sklearn.metrics import classification_report
print(classification_report(Y,y_predict))

              precision    recall  f1-score   support

           0       0.97      0.99      0.98      5489
           1       0.99      0.97      0.98      5289

    accuracy                           0.98     10778
   macro avg       0.98      0.98      0.98     10778
weighted avg       0.98      0.98      0.98     10778



# F score is 0.98 so our model is ready to predict if client witll do term deposit at bank or not